### DISCUSSION TITLE
同一routeで複数の検索条件に対応するには？

### DISCUSSION THREAD

**Discussion by tenk-9** [gdsc-tmu/InfoLinkAPI#28] [General]
https://github.com/gdsc-tmu/InfoLinkAPI/pull/25#discussion_r1507484878
ここから．
実際に使うときのことを考えると，シラバスの検索は複数の条件を組み合わせて使うことが多そう．複数条件での絞り込みに対応するとなると，
- 絞り込み条件ごとにrouteを分けるのは非効率
  - 複数条件での検索ができない
  - 無理やり全ての組み合わせについてrouteとcontrollerを定義すれば行けるけど，これは厳しい

なので，
- 検索のrouteを定義する
  - たとえば`syllabus/search`とか
- 絞り込み条件は検索routeのクエリパラメータで与える

のが良いんじゃないかと思います．

**Comment by tenk-9** [comment]
クエリパラメータで絞り込み条件を指定するとき
- おそらくrouter.GET()でcontrollerを呼び出す時点ではどんなクエリパラメータがあるかどうか分からない
  - クエリパラメータ（絞り込む条件）ごとに呼び出すcontrollerを変えるのはできない？

と思われるので，
- 検索routeで複数条件の絞り込みを担うcontrollerを呼び出すようにする
  - パラメータが複数あるとはいえ，route自体は同一なので

のが良いんじゃないでしょうか．

複数条件の絞り込みの実装については，
- 単一の条件について絞り込みを行う関数を定義しておく
- クエリパラメータと呼び出す絞り込み関数の対応付けをしておく
- クエリパラメータから絞り込み条件を抽出する
  - `c.Request.URL.Query()`でリクエストURL内のパラメータをmapで取得できるらしい（https://stackoverflow.com/questions/41279297/how-to-get-all-query-parameters-from-go-gin-context-object）
  - `c.Query("hoge")`で，指定したクエリパラメータの値を得ることもできる
    - URL内で指定されていなければ空文字列が返る
    - `syllabus/search?hoge=fuga`のとき，`c.Query("hoge")` returns `fuga`.
- リクエストURLで指定された絞り込み条件に対応づいた関数で絞り込みをする
- はじめに呼び出される絞り込み関数にはDB全体を渡し，次以降は絞り込み結果を渡せばいいんじゃないかな
  - 段階的にろ過されていくイメージ

のようにすればできそうだなと思っています，これでどうすかね？

**Comment by Atotti** [comment]
大体理解しました。これなら実装出来そうですが、パフォーマンスの面をもう少し考えたい気もします。

[IMO] こういうものは独自の実装をするメリットはほとんどなくて、既存の他のライブラリやサービスがどんな実装をしているか調べた方が良さそう。

**Reply by tenk-9** [reply]
たしかに，ちょっと調べてみます

**Reply by Atotti** [reply]
こちらでも少し調べてみますね
